index.rst

```diff
@@ -42,20 +42,163 @@
+The purpose is to provide early feedback to observing in the form of data quality assessments.
+
+.. [*] In :cite:`LTS-836` the group ID is called a "transaction ID" since it is associated with a single queue transaction.
+
+.. note::
+
+  No LSST rapid analysis pipelines are required to co-add data incrementally.
+  Pipelines should not co-add data as it arrives and should always defer co-addition until the final observation in the group has been taken.
+
+Controlling Data Processing from Observing Scripts
+==================================================
+
+In the initial design of the OCS Controlled Pipeline System (OCPS) [*]_ the pipeline processing will be controlled directly by commands sent from the observing script to a CSC.
+The commands will indicate which data to process and the name of the pipeline to execute but they will not include algorithmic code.
+The data will be specified using a "dataId" to match standard Data Butler usage.
+Once the pipeline completes an event will be published.
+
+.. [*] In :cite:`LDM-148` this is called the "OCS-Controlled Batch Service."
+
+The sequence of events for a typical observing script would then be:
+
+* Send a command to the camera to take an image.
+  This command should include the relevant groupId string as supplied from the ScriptQueue, and the observing type (such as science, bias, or flat).
+* Wait for the event indicating that the exposure has been written to the OODS (see :cite:`DMTN-111`).
+  This event will include the relevant obsid.
+* Send a command to the OCPS requesting that this obsid be processed.
+  This command should include the name of the pipeline to execute.
+* Acquire more data and send processing requests for each one.
+* When all data are acquired and single-frame processing has completed, send a command to the OCPS requesting that a co-add pipeline be executed with all data using the shared groupId.
+* An observing script can finish as soon as it has sent the final co-add command to the OCPS.
+
+In some cases a complex script might require multiple groups to be defined during execution.
+To deal with this case Python observing scripts shall be given an API to generate a new unique group ID from the value supplied to the script from the ScriptQueue.
+
+.. warning::
+
+   Observing scripts for standard visits and alternate standard visits should never use a group ID that differs from that supplied by the ScriptQueue.
+   This is because the prompt processing system uses the group ID to collate data and expects a specific value to appear that matches the value published by the ScriptQueue.
+
+As pipelines execute, they will be able to calculate data quality assessment metrics and other results of interest such as centroids.
+These can be published as telemetry to allow them to be displayed to the observer.
```

Comment:
In general we will need to display images too (e.g. master calibration frames).  Is it envisaged that this will be done via a callback to the command script (in which case the "gather" step cannot be run in the background), or are you thinking of some other mechanism for this sort of thing?

Reply:
I was imagining the display would be decoupled from the processing. A visualization system could listen for events relating to certain datasetTypes being stored in the butler repository and then decided whether it wants to display them or not. If the pipeline is running on the diagnostic cluster we would generate the JPEGs or whatever is needed .